torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -763,27 +773,57 @@ void Reducer::initialize_buckets(
 // (see Note:  "Gradient Layout Contract" in initialize_buckets).
 void Reducer::initialize_bucketviews(
     Reducer::BucketReplica& replica,
-    at::Tensor& contents) {
+    at::Tensor& contents,
+    bool copy_to_bucket_view) {
   for (size_t i = 0; i < replica.variables.size(); i++) {
-    const auto& v = replica.variables[i];
+    auto& v = replica.variables[i];
     const auto offset = replica.offsets[i];
     const auto length = replica.lengths[i];
+    at::Tensor bucket_view;
     if (v.is_non_overlapping_and_dense()) {
       // If the param's memory is dense, match its layout, anticipating
       // the autograd engine (AccumulateGrad) will also create gradients
       // matching its layout.
-      replica.bucket_views.push_back(
-          contents.as_strided(v.sizes(), v.strides(), offset));
+      bucket_view = contents.as_strided(v.sizes(), v.strides(), offset);
     } else {
       // Fall back to a C-style contiguous view, again anticipating
       // AccumulateGrad will do the same when stashing grads for non-dense
       // params.
-      replica.bucket_views.push_back(
-          contents.narrow(0, offset, length).view(v.sizes()));
+      bucket_view = contents.narrow(0, offset, length).view(v.sizes());
     }
+    replica.bucket_views.push_back(bucket_view);
+    // There are three cases to handle:
+    // 1. initialize_bucketviews could be called inside communication hook,
+    // bucket_view has the updated results in new tensor, just let grad point to
+    // bucket_view.
+    // 2. initialize_bucketviews could be called inside initialize_bucket when
+    // rebuildBuckets, if grad has already been defined/calculated in previous
+    // iteration, old grad needs to be copied into new bucket_view
+    // and let grad point to the new bucket_view;
+    // 3. initialize_bucketviews could be called inside initialize_bucket
+    // during construction. When grad is not defined, do not let it point to
+    // bucket_view, because grads should be kept as being undefined for globally
+    // unused parameters.
+    runGradCallbackForVariable(v, [&](auto& grad) {
+      if (grad.defined() && !grad.is_alias_of(bucket_view)) {
+        if (copy_to_bucket_view) {
+          bucket_view.copy_(grad);
+        }
+        grad = bucket_view;
+        // The grad is modefied and needs to be written back.
+        return true;
+      }
+      // The grad is not modified and does not need to be written back.
+      return false;
+    });
   }
 }
 
+void Reducer::prepare_forward() {
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
what race we are trying to protect here? I think this is the same discussion as in the Reducer ctor. I might miss sth, but it doesn't seem like there will be different thread mutating states here?

Reply:
prepare_forward is exposed to be public function of reducer, generally it needs to protect variables if multiple threads call this function? also, it is possible multiple threads calls forwards(), right?

this is similar to prepare_for_backward() where lock is added as well

Reply:
> prepare_forward is exposed to be public function of reducer, generally it needs to protect variables if multiple threads call this function? also, it is possible multiple threads calls forwards(), right?

I see, yep, the lock is also there for `prepare_for_backward`.